rye/src/cli/toolchain.rs

```diff
@@ -121,15 +144,28 @@ fn list(cmd: ListCommand) -> Result<(), Error> {
     let mut versions = toolchains.into_iter().collect::<Vec<_>>();
     versions.sort_by_cached_key(|a| (a.1.is_none(), a.0.kind.to_string(), Reverse(a.clone())));
 
-    for (version, path) in versions {
-        if let Some(path) = path {
-            println!(
-                "{} ({})",
-                style(&version).green(),
-                style(path.display()).dim()
-            );
-        } else {
-            println!("{} (downloadable)", style(version).dim());
+    if let Some(Format::Json) = cmd.format {
+        let json_versions = versions
+            .into_iter()
+            .map(|(version, path)| ListVersion {
+                name: version,
+                downloadable: if path.is_none() { Some(true) } else { None },
+                path: path.map(|p| p.to_string_lossy().into_owned()),
```

Comment:
the lossy part is trouble, maybe there's an easy enough solution in using an urlescaped file:/// url

Reply:
We can easily do a file url from Rust, but the Python libs on the other side (virtualenv discovery mechanism) can't deal with non-str paths anyway, right now, so no need to spend time on this case (?).

Reply:
I agree there is not a lot of point dwelling on that. In fact I noticed some tooling has issues with many non basic ascii paths.